package.json

```diff
@@ -41,6 +41,7 @@
   },
   "dependencies": {
     "debug": "^4.0.0",
+    "fast-deepclone": "^1.0.1",
```

Comment:
Can you use https://github.com/davidmarkclements/rfdc instead? I prefer no native dependency on this module.

Reply:
Answering your question above, deep-cloning should be done as unsuspecting users might pass-by-reference internal settings set by another plugin. Take the following snippet into consideration:

first-plugin.js: 
```
const fp = require('fastify-plugin')

module.exports = fp((fastify, options) => {
    // Pseudo-code
    const establishConnection = opts => createConnection(opts)
    // Pseudo-code end
    fastify.decorate('conn_settings', options) 
    fastify.decorate('create_connection_1', establishConnection)
})
```
second-plugin.js:
```
const fp = require('fastify-plugin')

module.exports = fp((fastify, opts) => {
    opts.port = Number.isNaN(opts.port) ? Number(opts.port) : opts.port
    fastify.decorate('my_conn_2', establishConnection(opts))
})
```
```
fastify.register(require('./first-plugin'));
fastify.register(require('./second-plugin'), ()=>fastify.conn_settings);
fastify.ready(err => {
    fastify.create_connection_1() // Error: Expected opts.port to be string but got number
})
```

I know this is not the most ideal example but it can become a danger nonetheless. This is an end-user error so we can ignore this if we wish so.

Sorry for the late responses, today is a busy day.

Reply:
The issue above is also indicated in the test sub-case on basic.test.js. I can revert the changes and simply write the initial `optsOrFunc instanceof Function ? optsOrFunc() : optsOrFunc` without the extras if you believe the deep-copy is unnecessary.

Reply:
Agree with @mcollina that we should not use native dependencies.
Regarding the deep cloning, I don't think we should introduce it, JavaScript passes objects by reference, and the user should know this.
If someone really needs to clone the object, they can do that by adding the cloning utility inside their function :)

Safe defaults for the win!

Reply:
I agree that removing deep cloning is the safest option here.

Reply:
I removed the additional test case (as it would fail) and removed the deep-cloning dependency. I should have created a basic commit and asked for enhancements rather than try to path up everything from the get go